tests/components/music_assistant/test_config_flow.py

```diff
@@ -0,0 +1,191 @@
+    )
+    config_entry.add_to_hass(hass)
+
+    await hass.config_entries.async_setup(config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    return config_entry
+
+
+async def test_full_flow(
+    hass: HomeAssistant,
+    mock_get_server_info,
+) -> None:
+    """Test full flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_USER},
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_URL: "http://localhost:8095"},
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Music Assistant"
+    assert result["data"] == {
+        CONF_URL: "http://localhost:8095",
+    }
+    assert result["result"].unique_id == "1234"
+
+
+async def test_zeor_conf_flow(
+    hass: HomeAssistant,
+    mock_get_server_info,
+) -> None:
+    """Test zeroconf flow."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_ZEROCONF},
+        data=ZEROCONF_DATA,
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "discovery_confirm"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {CONF_URL: "http://localhost:8095"},
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == "Music Assistant"
+    assert result["data"] == {
+        CONF_URL: "http://localhost:8095",
+    }
+    assert result["result"].unique_id == "1234"
+
+
+async def test_flow_user_init_connect_issue(
+    mock_get_server_info, hass: HomeAssistant
+) -> None:
+    """Test we advance to the next step when server url is invalid."""
+    mock_get_server_info.side_effect = CannotConnect("cannot_connect")
+
+    _result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": "user"}
+    )
+    result = await hass.config_entries.flow.async_configure(
+        _result["flow_id"], user_input={CONF_URL: "bad"}
+    )
+    assert result["errors"] == {"base": "cannot_connect"}
+
+
+async def test_flow_user_init_server_version_invalid(
+    mock_get_server_info, hass: HomeAssistant
+) -> None:
+    """Test we advance to the next step when server url is invalid."""
+    mock_get_server_info.side_effect = InvalidServerVersion("invalid_server_version")
+
+    _result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": "user"}
+    )
+    result = await hass.config_entries.flow.async_configure(
+        _result["flow_id"], user_input={CONF_URL: "bad"}
+    )
+    assert result["errors"] == {"base": "invalid_server_version"}
```

Comment:
Every config entry test has to end in either ABORT or CREATE_ENTRY to test it is able to recover from an exception. So we should repatch the function to work correctly and finish theflow.

Also, these tests can be deduplicated using `pytest.mark.parametrize`